code/parachain/frame/cosmwasm/src/ibc.rs

```diff
@@ -25,22 +59,325 @@ impl<T: Config> Pallet<T> {
+		cosmwasm_call_serialize::<IbcChannelOpen, WasmiVM<CosmwasmVM<T>>, IbcChannelOpenMsg>(
+			&mut executor,
+			&message,
+		)
+		.map_err(|err| IbcError::implementation_specific(format!("{:?}", err)))
+		.map(|x| match x.0 {
+			ContractResult::Ok(_) => Ok(()),
+			ContractResult::Err(err) =>
+				Err(IbcError::implementation_specific(format!("{:?}", err))),
+		})??;
+		let _remaining = vm.gas.remaining();
+		Ok(())
+	}
+
+	fn on_chan_open_try(
+		&mut self,
+		_output: &mut ModuleOutputBuilder,
+		order: Order,
+		connection_hops: &[ConnectionId],
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		counterparty: &Counterparty,
+		version: &IbcVersion,
+		counterparty_version: &IbcVersion,
+	) -> Result<IbcVersion, IbcError> {
+		let order = map_order(order);
+		let address = Self::port_to_address(port_id)?;
+		let contract_info = Self::to_ibc_contract(&address)?;
+
+		let message = {
+			use cosmwasm_minimal_std::ibc::{IbcChannel, IbcChannelOpenMsg, IbcEndpoint, IbcOrder};
+			IbcChannelOpenMsg::OpenInit {
+				channel: IbcChannel {
+					endpoint: IbcEndpoint {
+						channel_id: channel_id.to_string(),
+						port_id: port_id.to_string(),
+					},
+					counterparty_endpoint: IbcEndpoint {
+						port_id: counterparty.port_id.to_string(),
+						channel_id: counterparty.channel_id.expect("channel").to_string(),
+					},
+					order,
+					version: version.to_string(),
+					connection_id: connection_hops[0].to_string(),
+				},
+			}
+		};
+
+		let gas = Weight::MAX;
+		let mut vm = <Pallet<T>>::do_create_vm_shared(gas, InitialStorageMutability::ReadWrite);
+		let mut executor = Self::relayer_executor(&mut vm, address, contract_info)?;
+		let result = cosmwasm_call_serialize::<
+			IbcChannelOpen,
+			WasmiVM<CosmwasmVM<T>>,
+			IbcChannelOpenMsg,
+		>(&mut executor, &message)
+		.map_err(|err| IbcError::implementation_specific(format!("{:?}", err)))
+		.map(|x| match x.0 {
+			ContractResult::Ok(version) => Ok(version),
+			ContractResult::Err(err) =>
+				Err(IbcError::implementation_specific(format!("{:?}", err))),
+		})??
+		.map(|x| IbcVersion::new(x.version.to_string()))
+		.unwrap_or(version.clone());
+		let _remaining = vm.gas.remaining();
+		Ok(result)
+	}
+
+	fn on_chan_open_ack(
+		&mut self,
+		_output: &mut ModuleOutputBuilder,
+		_port_id: &PortId,
+		_channel_id: &ChannelId,
+		_counterparty_version: &IbcVersion,
+	) -> Result<(), IbcError> {
+		Ok(())
+	}
+
+	fn on_chan_open_confirm(
+		&mut self,
+		_output: &mut ModuleOutputBuilder,
+		_port_id: &PortId,
+		_channel_id: &ChannelId,
+	) -> Result<(), IbcError> {
+		Ok(())
+	}
+
+	fn on_chan_close_init(
+		&mut self,
+		_output: &mut ModuleOutputBuilder,
+		_port_id: &PortId,
+		_channel_id: &ChannelId,
+	) -> Result<(), IbcError> {
+		Ok(())
+	}
+
+	fn on_chan_close_confirm(
+		&mut self,
+		_output: &mut ModuleOutputBuilder,
+		_port_id: &PortId,
+		_channel_id: &ChannelId,
+	) -> Result<(), IbcError> {
+		Ok(())
```

Comment:
Similarly to OpenInit/OpenTry, we must call the contract IbcChannelCloseInput here